internal/venice-client-common/src/main/java/com/linkedin/venice/compression/ZstdWithDictCompressor.java

```diff
@@ -3,32 +3,40 @@
 import static com.linkedin.venice.utils.ByteUtils.BYTES_PER_KB;
 import static com.linkedin.venice.utils.ByteUtils.BYTES_PER_MB;
 
+import com.github.luben.zstd.Zstd;
 import com.github.luben.zstd.ZstdCompressCtx;
+import com.github.luben.zstd.ZstdDecompressCtx;
+import com.github.luben.zstd.ZstdDictCompress;
+import com.github.luben.zstd.ZstdDictDecompress;
 import com.github.luben.zstd.ZstdDictTrainer;
+import com.github.luben.zstd.ZstdException;
 import com.github.luben.zstd.ZstdInputStream;
 import com.linkedin.venice.compression.protocol.FakeCompressingSchema;
 import com.linkedin.venice.serializer.AvroSerializer;
-import com.linkedin.venice.utils.ByteUtils;
 import com.linkedin.venice.utils.concurrent.CloseableThreadLocal;
-import java.io.ByteArrayInputStream;
+import com.linkedin.venice.utils.lazy.Lazy;
 import java.io.IOException;
 import java.io.InputStream;
 import java.nio.ByteBuffer;
 import java.util.ArrayList;
 import java.util.List;
 import org.apache.avro.generic.GenericData;
 import org.apache.avro.generic.GenericRecord;
-import org.apache.commons.io.IOUtils;
 
 
 public class ZstdWithDictCompressor extends VeniceCompressor {
-  private CloseableThreadLocal<ZstdCompressCtx> compressor;
-  private byte[] dictionary;
+  private final CloseableThreadLocal<ZstdCompressCtx> compressor;
+  private final CloseableThreadLocal<ZstdDecompressCtx> decompressor;
+  private final Lazy<ZstdDictCompress> zstdDictCompress;
+  private final Lazy<ZstdDictDecompress> zstdDictDecompress;
 
   public ZstdWithDictCompressor(final byte[] dictionary, int level) {
     super(CompressionStrategy.ZSTD_WITH_DICT);
-    this.dictionary = dictionary;
-    compressor = new CloseableThreadLocal(() -> new ZstdCompressCtx().loadDict(dictionary).setLevel(level));
+    this.zstdDictCompress = Lazy.of(() -> new ZstdDictCompress(dictionary, level));
```

Comment:
The intent is that we have a class which does both compression and decompression, but some code paths need just one of these two. For example, the router only decompresses, and VPJ only compresses. So I was thinking we might want to avoid creating resources we won't need. If you prefer, I can create both always and make them non-lazy, which isn't a big deal IMHO.

Reply:
Conceptually, there might not be a big difference.
In read path, right after version switch, there will be many requests hitting the new version, and since LazyImpl is using a synchronized method to construct an instance, so all the decompression requests will wait for it, which may cause some latency spike.

I think it is good to not use Lazy for decompression, and it seems to be fine to use Lazy for compression, which is only used in write path.

Reply:
Not using lazy at all sounds good to me.

Reply:
Done.